Remove manual focus tracking for navigation link blocks

jeryj opened the pull request:
## What? 
Fixes and simplifies focus management issues in Navigation Link and Navigation Submenu blocks when closing link popovers. Fixes https://github.com/WordPress/gutenberg/issues/68035

Fixes https://github.com/WordPress/gutenberg/issues/65962 by moving block selection to the parent block so the sidebar editing isn't available during this step.

Flaky test fixes: Closes <!-- #ISSUE-NUMBER or URL --> https://github.com/WordPress/gutenberg/issues/71966, https://github.com/WordPress/gutenberg/issues/72064, https://github.com/WordPress/gutenberg/issues/72065 and https://github.com/WordPress/gutenberg/issues/71108


## Why? 
After adding a new navigation link and clicking anywhere (inspector, header, etc.), the popover would aggressively steal focus back to the canvas, preventing interaction with other UI elements, instead of letting the click naturally place focus. 

## How? 
Removed the openedBy state that was tracking which element opened the link popover, as it's no longer needed for focus management since https://github.com/WordPress/gutenberg/pull/68060 and https://github.com/WordPress/gutenberg/pull/71252/ were merged.

- Keep selection on parent block
- This allows the appender that opened the link ui to remain visible
- Modified onClose handler to: 
    - Only handle focus when adding a new link by selecting the newly added block.
    - If the link does not get created, remove the empty link
    - Focus is naturally handled by the useFocusOutside on the popover
- Broke navigation tests into steps and updated them to have focus go to the appender instead of previous block
- Added polling to fix flakey link ui check

Additional Changes
 - Applied the same fixes to both navigation-link and navigation-submenu blocks for consistency
 - Fixed React hook linting

Test Focus Behavior for New Links 
1. Add a Navigation block 
2. Click the appender to add a new Navigation Link (popover opens) 
3. Type a URL and press Enter (link is created) 
4. Click on an inspector control field in the sidebar 
5. ✅ Focus should stay in the inspector field you clicked 


Test Escape Key Behavior for New Links 
1. Add a Navigation block 
2. Click the appender to add a new Navigation Link (popover opens) 
3. Press Escape without entering a URL 
4. ✅ The empty block should be removed 
5. ✅ The previous block should be selected 
6. ✅ The appender should remain visible

Test Escape Key Behavior for Existing Links 
- From rich text element, press command + k to open the popover
- Press Escape
- ✅  Focus should go back to the rich text field
- Shift + tab to the toolbar
- Tab to the link edit button
- Enter
- Escape
- ✅ Focus should be on the toolbar link edit button

Test Navigation Submenu Repeat all the above tests with Navigation Submenu blocks to ensure they behave identically.

**Before**

https://github.com/user-attachments/assets/63c67475-6ce9-4e0c-953e-93d3a8edd697



**After**

https://github.com/user-attachments/assets/87d400ba-4277-46e9-9638-18102cf283d8

jeryj committed:
Remove usage of openedBy and rely on baseline tooling. Only handle focus on escape keypress

jeryj committed:
Only handle moving focus to new link if it's a new link

jeryj committed:
Apply changes to navigation-submenu also

jeryj committed:
Determine isNewLink by if it's selected on mount

jeryj committed:
Remove isNewLink tracking. Send focus back to appender after new link created

getdave reviewed (commented):
https://github.com/user-attachments/assets/7afc4f64-5493-49b0-8b4c-1f120147b749

Great to be fixing this. Unfortunately the bit about toolbar focus doesn't seem to be working (I think that's confirmed by the failing e2e tests as well). See video above.

jeryj commented:
@getdave Whoops - I hadn't pushed the rest of my local work. Friday afternoon brain :) I know there will be some e2e test changes to make based on the changes in this branch.

jeryj committed:
Select parent block when creatign new navigation link

When we're in the link creation flow, the block currently being created was selecting, creating a very unstable state.
- If you click off this modal, the link is removed
- In the sidebar inspector, the navigation link fields are visible for editing. clicking one of those removes your link, making it impossible to use these visible fields.
- If you escape out of the flow, your currently selected block is removed, causing a focus loss.

This PR selects the parent navigation block during this unstable link creation phase, so the sidebar shows the parent navigation block state and if you bail on the creation process, it doesn't cause a focus loss. It allows returning focus to the appender by default.

jeryj committed:
Select parent submenu block if navigation link is within a submenu when creating a link

jeryj committed:
If a new link is created, select the new link block after creation

jeryj committed:
Remove unnused escape key tracking

jeryj committed:
Update test to use steps

jeryj committed:
Fix submenu navigation tests

jeryj commented on packages/block-library/src/navigation-submenu/edit.js:
```diff
@@ -90,7 +90,7 @@ const useIsDraggingWithin = ( elementRef ) => {
 			ownerDocument.removeEventListener( 'dragenter', handleDragEnter );
 		};
-	}, [] );
+	}, [ elementRef ] );
```
Linting

getdave replied:
https://github.com/WordPress/gutenberg/pull/72098

jeryj commented on packages/block-library/src/navigation-submenu/edit.js:
```diff
@@ -324,10 +346,10 @@ export default function NavigationSubmenuEdit( {
 		const newLinkBlock = createBlock( 'core/navigation-link', attributes );
 		replaceBlock( clientId, newLinkBlock );
-	}
+	}, [ attributes, replaceBlock, clientId ] );
```
Linting

getdave replied:
https://github.com/WordPress/gutenberg/pull/72098

jeryj commented on packages/block-library/src/navigation-submenu/edit.js:
```diff
@@ -232,7 +255,7 @@ export default function NavigationSubmenuEdit( {
 			}
 		}
-	}, [ url ] );
+	}, [ url, isLinkOpen, label ] );
```
Linting

getdave replied:
https://github.com/WordPress/gutenberg/pull/72098

jeryj committed:
Cleanup

jeryj committed:
Remove linting fixes for submenus as those are being resolved by #72098

jeryj commented:
Closing in favor of https://github.com/WordPress/gutenberg/pull/72142, as the commits are cleaner
